pkg/acs/calacs/acsccd/blev_funcs_postsm4.c

```diff
@@ -81,102 +117,137 @@ int bias_shift_corr(ACSInfo *acs, SingleGroup *chip2, SingleGroup *chip1) {
   double sum;
   int num;
   double magic_square_mean;
-  
+
   /* allocate space for data arrays */
   ampdata = malloc(arr_rows * arr_cols * sizeof(double));
   ampdata_gap = malloc(nquad_pix * sizeof(double));
   
   /* allocate space for true DC bias levels array */
   dc_bias_levels = malloc((nquad_pix + 1) * sizeof(double));
-  
-  for (i = 0; i < NAMPS; i++) {
-    /* put a single amp's data into ampdata */
-    if (i < 2) {
-      make_amp_array(arr_rows, arr_cols, chip1, i, ampdata);
-    } else {
-      make_amp_array(arr_rows, arr_cols, chip2, i, ampdata);
-    }
+
+  /* For full frame data, there are four amps in use.  For supported subarray
+   * data, there is only one amp in use.  The actual amp in use must be determined. 
+   */
+  int numAmpsInUse = (acs->subarray == YES) ? 1 : NAMPS;
+  
+  /* Loop over all the amps for full frame data (total of 4 amps, 0 to 3) or just use 
+   * the single amp value corresponding to the supported subarray (1 amp, value defined in acs.h).
+   */
+  int ampInUse = 0;
+  for (i = 0; i < numAmpsInUse; i++) {
+      ampInUse = i;
+      if ((i == 0) && (acs->subarray == YES)) {
+         if (strcmp (acs->ccdamp, "A") == 0)
+            ampInUse = AMP_A;
+         else if (strcmp (acs->ccdamp, "B") == 0)
+            ampInUse = AMP_B;
+         else if (strcmp (acs->ccdamp, "C") == 0)
+            ampInUse = AMP_C;
+         else if (strcmp (acs->ccdamp, "D") == 0)
+            ampInUse = AMP_D;
+         else
+		    return (status = 1);
+      } 
+sprintf (MsgText, "Amp in use: %d  NumOfAmps: %d\n", ampInUse, numAmpsInUse);
+trlmessage(MsgText);
+
+      /* put a single amp's data into ampdata */
+      if (acs->subarray == YES) {
+         make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+      } else {
+         if (i < 2) {
+            make_amp_array(arr_rows, arr_cols, sg[1], ampInUse, ampdata);
+         } else {
+            make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+         }
+      }
        
-    /* calculate "magic square mean" */
-    sum = 0.0;
-    num = 0;
+      /* calculate "magic square mean" - this works for any single amp */
+      sum = 0.0;
+      num = 0;
     
-    for (j = 2057; j <= 2066; j++) {
-      for (k = 13; k <= 22; k++) {
-        sum += ampdata[arr_cols*j + k];
-        num++;
+      for (j = 2057; j <= 2066; j++) {
```

Comment:
Even prior to this PR, I was uncomfortable with these magic numbers. Perhaps it is time to make them constant variables with meaningful names?

This comment also applies to other occurrences of magic numbers in this file.

Reply:
This is already in progress. An updated overscan table has been created with primary header keywords and a new binary table in the second extension to accommodate amp-dependent values.